overlord/devicestate/devicestate.go

```diff
@@ -1186,6 +1193,28 @@ func checkForInvalidSnapsInModel(model *asserts.Model, vSets *snapasserts.Valida
 	return nil
 }
 
+func checkForSystemSeed(st *state.State, deviceCtx snapstate.DeviceContext) (bool, error) {
+	// on non-classic systems, we will always have a seed partition
+	if !deviceCtx.Classic() {
+		return true, nil
+	}
+
+	gadgetData, err := CurrentGadgetData(st, deviceCtx)
+	if err != nil {
+		return false, fmt.Errorf("cannot get gadget data: %w", err)
+	}
+
+	for _, vol := range gadgetData.Info.Volumes {
+		for _, vs := range vol.Structure {
+			if vs.Role == gadget.SystemSeed {
+				return true, nil
+			}
+		}
+	}
```

Comment:
maybe this should be a method on gadget.Info? see related but not quite useful here gadget.HasRole